app/scripts/modules/core/src/api/ApiService.ts

```diff
@@ -1,193 +1,254 @@
 
 type IPrimitive = string | boolean | number;
 type IParams = Record<string, IPrimitive | IPrimitive[]>;
 
+/**
+ * A Builder API for making requests to Gate backend service
+ */
 export interface IRequestBuilder {
-  config?: IRequestConfig;
-  one?: (...urls: string[]) => IRequestBuilder;
-  all?: (...urls: string[]) => IRequestBuilder;
-  useCache?: (useCache?: boolean | ICache) => IRequestBuilder;
-  withParams?: (params: IParams) => IRequestBuilder;
-  data?: (data: any) => IRequestBuilder;
-  get?: <T = any>(params?: IParams) => PromiseLike<T>;
-  getList?: <T = any>(params?: IParams) => PromiseLike<T>;
-  post?: <T = any>(data?: any) => PromiseLike<T>;
-  remove?: <T = any>(params?: IParams) => PromiseLike<T>;
-  put?: <T = any>(data?: any) => PromiseLike<T>;
+  /**
+   * Appends one or more path segments to the URL, separated by slashes.
+   * Each path segment is uri encoded.
+   */
+  path(...pathSegments: IPrimitive[]): this;
+
+  /** Adds query parameters to the URL */
+  query(queryParams: IParams): this;
+
+  /** Enables or disables caching of the response */
+  useCache(useCache?: boolean): this;
+
+  /** issues a GET request */
+  get<T = any>(): PromiseLike<T>;
+  /** issues a POST request */
+  post<T = any, P = any>(data?: P): PromiseLike<T>;
+  /** issues a PUT request */
+  put<T = any, P = any>(data?: P): PromiseLike<T>;
+  /** issues a DELETE request */
+  delete<T = any>(): PromiseLike<T>;
+}
+
+/**
+ * Internal interface to encapsulate a request
+ * Passed to IHttpClientBackend
+ */
+interface IRequestBuilderConfig {
+  url: string;
+  timeout?: number;
+  headers?: { [headerName: string]: string };
+  /** @deprecated used for AngularJS backwards compat */
+  data?: any;
+  params?: object;
+  cache?: boolean;
+}
+
+/**
+ * The old API interface
+ */
+export interface IDeprecatedRequestBuilder extends IRequestBuilder {
+  useCache(): this;
+  useCache(useCache: boolean): this;
+  useCache(useCache: ICache): this;
+  withParams(queryParams: IParams): this;
+
+  /** @deprecated do not use this config object */
+  config: IRequestBuilderConfig;
+  /** @deprecated use SETTINGS.gateUrl */
+  baseUrl: string;
+  /** @deprecated use path() instead (this is a passthrough to path) */
+  one(...urls: string[]): this;
+  /** @deprecated use one() instead (this is a passthrough to one) */
+  all(...urls: string[]): this;
+  /** @deprecated use put(data) or post(data) instead */
+  data(data: any): this;
+  // Add overload with params
+  get<T = any>(params?: IParams): PromiseLike<T>;
+  /** @deprecated use delete() instead (this is a passthrough to delete) */
+  remove(params?: IParams): PromiseLike<any>;
+  /** @deprecated use get() instead (this is a passthrough to get) */
+  getList<T = any>(params?: IParams): PromiseLike<T>;
+}
+
+/**
+ * An interface to support pluggable http clients
+ * In the future, we should have a TestingHttpBackend and a FetchHttpBackend (or whatever http client we go with)
```

Comment:
Might just be because I haven't done too much Angular, but I find the term "backend" confusing when we're really just referring to an http client.

Now might not be the best time, but can these (eventually) just be referred to as `HttpClient`s? Or is there a distinction I'm missing?

Reply:
I think there should be a way to distinguish between `REST()` being the HTTP client and `$http` being the specific "implementation".  ideas?

Reply:
May be just using the word "implementation"?`TestingHttpImplementation`, `FetchHttpImplementation` sound better to me.

Reply:
+1 ^^

Reply:
sounds good to me